dataflowengineoss/src/main/scala/io/joern/dataflowengineoss/passes/reachingdef/DdgGenerator.scala

```diff
@@ -172,10 +172,27 @@ class DdgGenerator(semantics: Semantics) {
       val identifierDestPairs =
         method._identifierViaContainsOut
           .flatMap { identifier =>
-            identifierToFirstUsages(identifier).map(usage => (identifier, usage))
+            val firstAndLastUsageByMethod = identifierToFirstUsages(identifier).groupBy(_.method)
+            firstAndLastUsageByMethod.values.flatMap { usage =>
+              if (usage.nonEmpty) {
```

Comment:
Perhaps instead of all these if-statements calls to `isDefined` we could use some filtering and pattern matching.
```scala
firstAndLastUsageByMethod
   .values
   .filter(_.nonEmpty)
   .map(x => (x.head, x.last))
   .flatMap { case (firstUsage, lastUsage) => 
    (identifier.lineNumber, firstUsage.lineNumber, lastUsage.lineNumber) match {
       case (Some(iNo), Some(fNo), _) if iNo <= fNo => Some(identifier, firstUsage)
       case (Some(iNo), _, Some(lNo)) if iNo >= lNo => Some(identifier, lastUsage)
       case _ => None
    }
 }

Reply:
Updated this to reflect the original ordering